File: packages/java/src/test/java/io/constellationnetwork/metagraph/sdk/IntegrationTest.java
package io.constellationnetwork.metagraph.sdk;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class IntegrationTest {

    @Nested
    @DisplayName("Key generation")
    class KeyGeneration {

        @Test
        @DisplayName("generates valid key pair")
        void generatesValidKeyPair() {
            Types.KeyPair keyPair = Wallet.generateKeyPair();

            assertEquals(64, keyPair.getPrivateKey().length());
            assertEquals(130, keyPair.getPublicKey().length());
            assertTrue(keyPair.getPublicKey().startsWith("04"));
            assertTrue(keyPair.getAddress().startsWith("DAG"));
        }

        @Test
        @DisplayName("derives consistent key pair")
        void derivesConsistentKeyPair() {
            Types.KeyPair original = Wallet.generateKeyPair();
            Types.KeyPair derived = Wallet.keyPairFromPrivateKey(original.getPrivateKey());

            assertEquals(original.getPublicKey(), derived.getPublicKey());
            assertEquals(original.getAddress(), derived.getAddress());
        }

        @Test
        @DisplayName("generates unique key pairs")
        void generatesUniqueKeyPairs() {
            Types.KeyPair keyPair1 = Wallet.generateKeyPair();
            Types.KeyPair keyPair2 = Wallet.generateKeyPair();

            assertNotEquals(keyPair1.getPrivateKey(), keyPair2.getPrivateKey());
            assertNotEquals(keyPair1.getPublicKey(), keyPair2.getPublicKey());
        }
    }

    @Nested
    @DisplayName("Regular signing")
    class RegularSigning {

        @Test
        @DisplayName("signs and verifies data")
        void signsAndVerifiesData() {
            Types.KeyPair keyPair = Wallet.generateKeyPair();
            Map<String, Object> data = new HashMap<>();
            data.put("id", "test-001");
            data.put("value", 42);

            Types.SignatureProof proof = Sign.sign(data, keyPair.getPrivateKey());

            assertNotNull(proof.getId());
            assertNotNull(proof.getSignature());
            assertEquals(128, proof.getId().length());

            boolean isValid = Verify.verifySignature(data, proof, false);
            assertTrue(isValid);
        }

        @Test
        @DisplayName("creates signed object")
        void createsSignedObject() {
            Types.KeyPair keyPair = Wallet.generateKeyPair();
            Map<String, Object> data = new HashMap<>();
            data.put("id", "test-001");
            data.put("value", 42);

            Types.Signed<Map<String, Object>> signed = SignedObject.createSignedObject(
                data, keyPair.getPrivateKey(), false
            );

            assertEquals(1, signed.getProofs().size());

            Types.VerificationResult result = SignedObject.verify(signed, false);
            assertTrue(result.isValid());
            assertEquals(1, result.getValidProofs().size());
            assertTrue(result.getInvalidProofs().isEmpty());
        }
    }

    @Nested
    @DisplayName("DataUpdate signing")
    class DataUpdateSigning {

        @Test
        @DisplayName("signs and verifies DataUpdate")
        void signsAndVerifiesDataUpdate() {
            Types.KeyPair keyPair = Wallet.generateKeyPair();
            Map<String, Object> data = new HashMap<>();
            data.put("id", "test-update-001");
            data.put("value", 123);

            Types.SignatureProof proof = Sign.signDataUpdate(data, keyPair.getPrivateKey());

            boolean isValid = Verify.verifySignature(data, proof, true);
            assertTrue(isValid);
        }

        @Test
        @DisplayName("verification fails with wrong mode")
        void verificationFailsWithWrongMode() {
            Types.KeyPair keyPair = Wallet.generateKeyPair();
            Map<String, Object> data = new HashMap<>();
            data.put("id", "test");
            data.put("value", 42);

            // Sign as DataUpdate
            Types.SignatureProof proof = Sign.signDataUpdate(data, keyPair.getPrivateKey());

            // Verify as regular (should fail)
            boolean isValid = Verify.verifySignature(data, proof, false);
            assertFalse(isValid);
        }

        @Test
        @DisplayName("produces different signatures than regular")
        void producesDifferentSignaturesThanRegular() {
            Types.KeyPair keyPair = Wallet.generateKeyPair();
            Map<String, Object> data = new HashMap<>();
            data.put("id", "test");
            data.put("value", 42);

            Types.SignatureProof regularProof = Sign.sign(data, keyPair.getPrivateKey());
            Types.SignatureProof dataUpdateProof = Sign.signDataUpdate(data, keyPair.getPrivateKey());

            assertNotEquals(regularProof.getSignature(), dataUpdateProof.getSignature());
        }
    }

    @Nested
    @DisplayName("Multi-signature")
    class MultiSignature {

        @Test
        @DisplayName("adds signature to existing object")
        void addsSignatureToExistingObject() {
            Types.KeyPair keyPair1 = Wallet.generateKeyPair();
            Types.KeyPair keyPair2 = Wallet.generateKeyPair();
            Map<String, Object> data = new HashMap<>();
            data.put("id", "multi-sign-test");
            data.put("value", 100);

            Types.Signed<Map<String, Object>> signed = SignedObject.createSignedObject(
                data, keyPair1.getPrivateKey(), false
            );
            Types.Signed<Map<String, Object>> multiSigned = SignedObject.addSignature(
                signed, keyPair2.getPrivateKey(), false
            );

            assertEquals(2, multiSigned.getProofs().size());

            Types.VerificationResult result = SignedObject.verify(multiSigned, false);
            assertTrue(result.isValid());
            assertEquals(2, result.getValidProofs().size());
        }

        @Test
        @DisplayName("batch signs with multiple keys")
        void batchSignsWithMultipleKeys() {
            Types.KeyPair keyPair1 = Wallet.generateKeyPair();
            Types.KeyPair keyPair2 = Wallet.generateKeyPair();
            Types.KeyPair keyPair3 = Wallet.generateKeyPair();
            Map<String, Object> data = new HashMap<>();
            data.put("id", "batch-sign-test");

            Types.Signed<Map<String, Object>> signed = SignedObject.batchSign(
                data,
                Arrays.asList(keyPair1.getPrivateKey(), keyPair2.getPrivateKey(), keyPair3.getPrivateKey()),
                false
            );

            assertEquals(3, signed.getProofs().size());

            Types.VerificationResult result = SignedObject.verify(signed, false);
            assertTrue(result.isValid());
            assertEquals(3, result.getValidProofs().size());
        }
    }

    @Nested
    @DisplayName("Tamper detection")
    class TamperDetection {

        @Test
        @DisplayName("detects modified value")
        void detectsModifiedValue() {
            Types.KeyPair keyPair = Wallet.generateKeyPair();
            Map<String, Object> originalData = new HashMap<>();
            originalData.put("id", "test");
            originalData.put("value", 42);

            Types.SignatureProof proof = Sign.sign(originalData, keyPair.getPrivateKey());

            // Tamper with data
            Map<String, Object> tamperedData = new HashMap<>();
            tamperedData.put("id", "test");
            tamperedData.put("value", 999);

            Types.Signed<Map<String, Object>> signed = new Types.Signed<>(
                tamperedData, Arrays.asList(proof)
            );

            Types.VerificationResult result = Verify.verify(signed, false);
            assertFalse(result.isValid());
            assertEquals(1, result.getInvalidProofs().size());
        }
    }

    @Nested
    @DisplayName("Hashing")
    class HashingTests {

        @Test
        @DisplayName("produces consistent hashes")
        void producesConsistentHashes() {
            Map<String, Object> data = new HashMap<>();
            data.put("id", "test");
            data.put("value", 42);

            Types.Hash hash1 = Hash.hash(data);
            Types.Hash hash2 = Hash.hash(data);

            assertEquals(hash1.getValue(), hash2.getValue());
        }

        @Test
        @DisplayName("hash is 32 bytes")
        void hashIs32Bytes() {
            Map<String, Object> data = new HashMap<>();
            data.put("id", "test");

            Types.Hash hash = Hash.hash(data);

            assertEquals(64, hash.getValue().length()); // 32 bytes = 64 hex chars
            assertEquals(32, hash.getBytes().length);
        }

        @Test
        @DisplayName("different data produces different hash")
        void differentDataProducesDifferentHash() {
            Map<String, Object> data1 = new HashMap<>();
            data1.put("id", "test1");

            Map<String, Object> data2 = new HashMap<>();
            data2.put("id", "test2");

            Types.Hash hash1 = Hash.hash(data1);
            Types.Hash hash2 = Hash.hash(data2);

            assertNotEquals(hash1.getValue(), hash2.getValue());
        }
    }

    @Nested
    @DisplayName("Error handling")
    class ErrorHandling {

        @Test
        @DisplayName("rejects invalid private key")
        void rejectsInvalidPrivateKey() {
            assertFalse(Wallet.isValidPrivateKey("invalid"));
            assertFalse(Wallet.isValidPrivateKey("abc123")); // Too short
            assertFalse(Wallet.isValidPrivateKey("zz" + "a".repeat(62))); // Invalid chars

            assertThrows(Types.SdkException.class, () ->
                Wallet.keyPairFromPrivateKey("invalid")
            );
        }

        @Test
        @DisplayName("batch sign requires at least one key")
        void batchSignRequiresAtLeastOneKey() {
            Map<String, Object> data = new HashMap<>();
            data.put("id", "test");

            assertThrows(Types.SdkException.class, () ->
                SignedObject.batchSign(data, Arrays.asList(), false)
            );
        }
    }
}

File: packages/go/integration_test.go
package constellation

import (
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestKeyGeneration(t *testing.T) {
	t.Run("generates valid key pair", func(t *testing.T) {
		keyPair, err := GenerateKeyPair()
		require.NoError(t, err)

		assert.Len(t, keyPair.PrivateKey, 64, "Private key should be 64 hex chars")
		assert.Len(t, keyPair.PublicKey, 130, "Public key should be 130 hex chars")
		assert.True(t, strings.HasPrefix(keyPair.Address, "DAG"), "Address should start with DAG")
	})

	t.Run("derives consistent key pair", func(t *testing.T) {
		original, err := GenerateKeyPair()
		require.NoError(t, err)

		derived, err := KeyPairFromPrivateKey(original.PrivateKey)
		require.NoError(t, err)

		assert.Equal(t, original.PublicKey, derived.PublicKey)
		assert.Equal(t, original.Address, derived.Address)
	})

	t.Run("generates unique key pairs", func(t *testing.T) {
		key1, err := GenerateKeyPair()
		require.NoError(t, err)

		key2, err := GenerateKeyPair()
		require.NoError(t, err)

		assert.NotEqual(t, key1.PrivateKey, key2.PrivateKey)
		assert.NotEqual(t, key1.PublicKey, key2.PublicKey)
		assert.NotEqual(t, key1.Address, key2.Address)
	})
}

func TestRegularSigning(t *testing.T) {
	t.Run("signs and verifies data", func(t *testing.T) {
		keyPair, err := GenerateKeyPair()
		require.NoError(t, err)

		data := map[string]interface{}{
			"action": "test",
			"value":  42,
		}

		signed, err := CreateSignedObject(data, keyPair.PrivateKey, false)
		require.NoError(t, err)

		result := Verify(signed, false)
		assert.True(t, result.IsValid)
		assert.Len(t, result.ValidProofs, 1)
		assert.Empty(t, result.InvalidProofs)
	})

	t.Run("signature contains public key id", func(t *testing.T) {
		keyPair, err := GenerateKeyPair()
		require.NoError(t, err)

		data := map[string]interface{}{"test": true}
		proof, err := Sign(data, keyPair.PrivateKey)
		require.NoError(t, err)

		// ID should be public key without 04 prefix
		assert.Len(t, proof.ID, 128)
		assert.Equal(t, keyPair.PublicKey[2:], proof.ID)
	})
}

func TestDataUpdateSigning(t *testing.T) {
	t.Run("signs and verifies data update", func(t *testing.T) {
		keyPair, err := GenerateKeyPair()
		require.NoError(t, err)

		data := map[string]interface{}{
			"id":    "update-001",
			"value": 123,
		}

		signed, err := CreateSignedObject(data, keyPair.PrivateKey, true)
		require.NoError(t, err)

		result := Verify(signed, true)
		assert.True(t, result.IsValid)
	})

	t.Run("data update verification fails with wrong mode", func(t *testing.T) {
		keyPair, err := GenerateKeyPair()
		require.NoError(t, err)

		data := map[string]interface{}{"id": "test"}

		// Sign as DataUpdate
		signed, err := CreateSignedObject(data, keyPair.PrivateKey, true)
		require.NoError(t, err)

		// Verify as regular (should fail)
		result := Verify(signed, false)
		assert.False(t, result.IsValid)
	})

	t.Run("produces different signatures than regular", func(t *testing.T) {
		keyPair, err := GenerateKeyPair()
		require.NoError(t, err)

		data := map[string]interface{}{"id": "test"}

		regularProof, err := Sign(data, keyPair.PrivateKey)
		require.NoError(t, err)

		updateProof, err := SignDataUpdate(data, keyPair.PrivateKey)
		require.NoError(t, err)

		// Same key
		assert.Equal(t, regularProof.ID, updateProof.ID)
		// Different signatures
		assert.NotEqual(t, regularProof.Signature, updateProof.Signature)
	})
}

func TestMultiSignature(t *testing.T) {
	t.Run("adds signature to existing object", func(t *testing.T) {
		key1, err := GenerateKeyPair()
		require.NoError(t, err)

		key2, err := GenerateKeyPair()
		require.NoError(t, err)

		data := map[string]interface{}{"action": "multi-sig"}

		signed, err := CreateSignedObject(data, key1.PrivateKey, false)
		require.NoError(t, err)

		signed, err = AddSignature(signed, key2.PrivateKey, false)
		require.NoError(t, err)

		assert.Len(t, signed.Proofs, 2)

		result := Verify(signed, false)
		assert.True(t, result.IsValid)
		assert.Len(t, result.ValidProofs, 2)
	})

	t.Run("batch signs with multiple keys", func(t *testing.T) {
		key1, err := GenerateKeyPair()
		require.NoError(t, err)

		key2, err := GenerateKeyPair()
		require.NoError(t, err)

		key3, err := GenerateKeyPair()
		require.NoError(t, err)

		data := map[string]interface{}{"action": "batch"}

		signed, err := BatchSign(data, []string{key1.PrivateKey, key2.PrivateKey, key3.PrivateKey}, false)
		require.NoError(t, err)

		assert.Len(t, signed.Proofs, 3)

		result := Verify(signed, false)
		assert.True(t, result.IsValid)
		assert.Len(t, result.ValidProofs, 3)
	})

	t.Run("all signatures are unique", func(t *testing.T) {
		key1, err := GenerateKeyPair()
		require.NoError(t, err)

		key2, err := GenerateKeyPair()
		require.NoError(t, err)

		data := map[string]interface{}{"id": "test"}

		signed, err := BatchSign(data, []string{key1.PrivateKey, key2.PrivateKey}, false)
		require.NoError(t, err)

		assert.NotEqual(t, signed.Proofs[0].ID, signed.Proofs[1].ID)
		assert.NotEqual(t, signed.Proofs[0].Signature, signed.Proofs[1].Signature)
	})
}

func TestTamperDetection(t *testing.T) {
	t.Run("detects modified value", func(t *testing.T) {
		keyPair, err := GenerateKeyPair()
		require.NoError(t, err)

		original := map[string]interface{}{"amount": 100}
		proof, err := Sign(original, keyPair.PrivateKey)
		require.NoError(t, err)

		// Create signed object with tampered value
		tampered := map[string]interface{}{"amount": 999}
		signed := &Signed[map[string]interface{}]{
			Value:  tampered,
			Proofs: []SignatureProof{*proof},
		}

		result := Verify(signed, false)
		assert.False(t, result.IsValid)
		assert.Empty(t, result.ValidProofs)
		assert.Len(t, result.InvalidProofs, 1)
	})

	t.Run("detects modified signature", func(t *testing.T) {
		keyPair, err := GenerateKeyPair()
		require.NoError(t, err)

		data := map[string]interface{}{"id": "test"}
		proof, err := Sign(data, keyPair.PrivateKey)
		require.NoError(t, err)

		// Tamper with signature
		proof.Signature = strings.ReplaceAll(proof.Signature, "0", "1")

		signed := &Signed[map[string]interface{}]{
			Value:  data,
			Proofs: []SignatureProof{*proof},
		}

		result := Verify(signed, false)
		assert.False(t, result.IsValid)
	})
}

func TestCodecOperations(t *testing.T) {
	t.Run("encodes and decodes data update", func(t *testing.T) {
		data := map[string]interface{}{
			"id":     "test",
			"nested": map[string]interface{}{"key": "value"},
			"array":  []int{1, 2, 3},
		}

		encoded, err := EncodeDataUpdate(data)
		require.NoError(t, err)

		var decoded map[string]interface{}
		err = DecodeDataUpdate(encoded, &decoded)
		require.NoError(t, err)

		assert.Equal(t, data["id"], decoded["id"])
	})

	t.Run("canonicalizes json consistently", func(t *testing.T) {
		data := map[string]interface{}{
			"z": 26,
			"a": 1,
			"m": 13,
		}

		canonical, err := Canonicalize(data)
		require.NoError(t, err)

		assert.Equal(t, `{"a":1,"m":13,"z":26}`, canonical)
	})

	t.Run("to bytes produces different output for data update", func(t *testing.T) {
		data := map[string]interface{}{"id": "test"}

		regular, err := ToBytes(data, false)
		require.NoError(t, err)

		update, err := ToBytes(data, true)
		require.NoError(t, err)

		assert.NotEqual(t, regular, update)
		assert.Contains(t, string(update), "Constellation Signed Data")
	})
}

func TestHashing(t *testing.T) {
	t.Run("produces consistent hashes", func(t *testing.T) {
		data := map[string]interface{}{"id": "test", "value": 42}

		hash1, err := HashData(data, false)
		require.NoError(t, err)

		hash2, err := HashData(data, false)
		require.NoError(t, err)

		assert.Equal(t, hash1.Value, hash2.Value)
		assert.Equal(t, hash1.Bytes, hash2.Bytes)
	})

	t.Run("hash is 32 bytes", func(t *testing.T) {
		data := map[string]interface{}{"test": true}

		hash, err := HashData(data, false)
		require.NoError(t, err)

		assert.Len(t, hash.Bytes, 32)
		assert.Len(t, hash.Value, 64)
	})

	t.Run("different data produces different hash", func(t *testing.T) {
		data1 := map[string]interface{}{"value": 1}
		data2 := map[string]interface{}{"value": 2}

		hash1, err := HashData(data1, false)
		require.NoError(t, err)

		hash2, err := HashData(data2, false)
		require.NoError(t, err)

		assert.NotEqual(t, hash1.Value, hash2.Value)
	})
}

func TestErrorHandling(t *testing.T) {
	t.Run("rejects invalid private key", func(t *testing.T) {
		_, err := KeyPairFromPrivateKey("invalid")
		assert.Error(t, err)
	})

	t.Run("batch sign requires at least one key", func(t *testing.T) {
		data := map[string]interface{}{"test": true}
		_, err := BatchSign(data, []string{}, false)
		assert.Equal(t, ErrNoPrivateKeys, err)
	})
}
